src/btree/bt_io.c

```diff
@@ -52,9 +52,22 @@ __wt_bt_read(WT_SESSION_IMPL *session,
 			    "configured");
 
 		/*
-		 * We're allocating the exact number of bytes we're expecting
+		 * Size the buffer based on the in-memory bytes we're expecting
 		 * from decompression.
+		 *
+		 * XXX
+		 * The underlying LZ4 raw compression API may have compressed
+		 * more data than we're actually going to use, in other words,
+		 * the final in-memory size of the decompressed data isn't big
+		 * enough and LZ4 decompression might fail where the destination
+		 * buffer is too small. The target size for LZ4 raw compression
+		 * is the maximum page size: start with the maximum page size
+		 * to be sure we allocate enough room, then call one more time
+		 * to allow the in-memory size to override and set the buffer's
+		 * final size to the in-memory size.
 		 */
+		WT_ERR(__wt_buf_initsize(session,
+		    buf, WT_MAX(btree->maxleafpage, btree->maxintlpage)));
 		WT_ERR(__wt_buf_initsize(session, buf, dsk->mem_size));
```

Comment:
This code is intended to handle a problem we have with the new API. In summary:

- WiredTiger now uses the raw compression code to interface to LZ4 compression, in other words, we put a bunch of rows into a buffer and compress as many of them as we can into a target size.
- LZ4's `LZ4_compress_destSize` API doesn't know anything about rows, it just compresses from the source buffer until nothing additional will fit into the target size.
- Because LZ4 doesn't stop on a row boundary, we have two values: the length of the data we compressed (which LZ4 needs to decompress), and the length of the useful data we compressed, which is what WiredTiger tracks as "the in-memory size of the uncompressed version of this block".
- WiredTiger uses that tracked size to allocate a buffer for decompression, and it might be too small, because LZ4 actually compressed more data than we are going to use.

Since the target for LZ4 compression is the maximum page size, I use the maximum of the btree leaf and internal page sizes and the in-memory size to size the buffer for decompression, which guarantees the target buffer is big enough.

The problem with this is the maximum btree leaf/internal page size might be pretty large and we're wasting a lot of space by doing this.

Alternative solutions:

1. Pass a `WT_ITEM` to the compression/decompression functions and let them grow the target buffer as necessary. We already pass WT_ITEM structures into the collator functions, I don't think this is insane, but the downside is it's a relatively big API change.
2. When doing LZ4 compression, make two calls: one to find out how much data we can compress into the buffer, and then one to do it again for real, that way the in-memory size is going to match the compressed size. Obvious downsize is doubling the number of compression calls we make.
3. Have LZ4 decompression allocate a scratch buffer if it needs more space than we give it and copy the results after decompression.

I'm inclined to go with solution 3 unless someone has a better idea.